multithread the MD5/IQM model CPU code using OpenMP

illwieckz opened the pull request:
- Add `Omp` facilities in framework
- Use OpenMP to multithread the MD5/IQM model CPU code

For now it is disabled by default, one should use the `-DUSE_OPENMP=ON` cmake option to enable it.

In the future I plan to progressively enable it:

- Expected: Enable it on Linux with GCC,
  we would have to modify the release validation script to accept the fact the executable depends on `libgomp.so`.
  The `libgomp.so` library is as standard as the glibc so it's fine.
- Probable: Enable it on Windows with MinGW, 
  we have to modify the release validation script to accept the fact the executable depends on `libgomp.dll`,
  and we would have to modify the release build script to package `libgomp.dll`.
  The `libgomp.dll` is provided by MSYS2 so it's fine.

I don't plan to enable it on macOS as I've heard that macOS doesn't ship the LLVM's `libomp.so` by default.

Such enablement will be done on later PRs.

The purpose of adding OpenMP abilities is to make it optional to speed-up operations with it, but the same operations should work without it.

Then it implements parallelization of the MD5 and IQM CPU code.

This was investigated on:

- https://github.com/DaemonEngine/Daemon/pull/1833
- https://github.com/DaemonEngine/Daemon/pull/1837

It uses a chunked implementation, as tests demonstrated it was the fastest one.

Using a beefy computer and enabling 16 threads I got that performance difference with the chunked implementation on the same heavy scene:

Before|After
-|-
`91fps`|`438fps`

Of course, the performance difference is expected to be lower on older CPUs usually running alongside older GPUs whose limitations enforce that CPU codepath, but it is now demonstrated that such parallelization scales well. This can move some devices from the `slow` to the `playable` category, or from the `playable` to the `passed` category.

A good way to test that is to follow those instructions:

```
/set r_vboModels off
/devmap plat23
/team h; class rifle; delay 1s setviewpos 1920 1920 20 0 0
```

This will spawn the human player and move it to the alien base entrance, where all the IQM buildable models from the alien base will be rendered because of them being in vis, with at least two IQM animated acid tubes actually in direct sight, plus the MD5 first person rifle on foreground. Starting from that one can also shot the acide tubes and empty the rifle magazines to play additional animations from the acid tube death and the rifle first person shoot and reload.

One can test various amount of threads this way:

```
/set common.ompThreads 4
```

The `0` default value let the engine picks an amount of threads by itself, other values enforce the amount of threads.

illwieckz commented:
For now there is a code that is guarded by a `NO_MT_IF_NO_TBNTOQ` define. This is because this code doesn't use `R_TBNtoQtangents()` and I want to test if that other code not using `R_TBNtoQtangents()` is slow enough to benefit from the parallelism (it looks like it is, but I will test more).

illwieckz commented:
> For now there is a code that is guarded by a `NO_MT_IF_NO_TBNTOQ` define. This is because this code doesn't use `R_TBNtoQtangents()` and I want to test if that other code not using `R_TBNtoQtangents()` is slow enough to benefit from the parallelism (it looks like it is, but I will test more).

On a machine with 8 cores and only running the game so the framerate is more stable, with the parallel code for that part I get `85 fps`, with the legacy code I get `80 fps`. That confirms what I have observed on my main machine, where framerate was much unstable due to other applications running around, and then doubt was allowed.

I'll drop the legacy sequential code for that part as well.

illwieckz commented:
While I was at it, I parallelized some parts of the MD3, MD5 and IQM loading code. The parallelization of the MD3 loading code is a bit noisy on the diff size because unlike the MD5 and IQM code that I cleaned-up long time ago, the MD3 code was full of reused “global to functions” variables that would just create race conditions once the code is parallelized.

illwieckz commented:
Now that I think about it, it's probably possible to template the chunking as well.

illwieckz commented:
> I bet the problem with your non-library-using chunked implementation was just that it spends too much time creating and destroying threads, which is solved by a thread pool.

Yes, this is a very bad solution for various problems (along with the fact that not reusing the threads makes profiling a nightmare).

> On the other hand, if you did take the OMP dependency, I imagine you would get somewhat better results with less code, by just putting the #pragma omp parallel for directly on the loops you want to parallelize, instead of dividing it into chunks.

I may have not tested yet the non-chunked code with the pragma indeed. I don't know what magic OMP does behind the scene when adding such pragma, but I found interesting that using the pragma on the chunked version was faster than the gnu parallel foreach call which did more guessing than the pragma on the for loop.

illwieckz commented:
> The latest version just uses OMP as a basic thread pool. You can find various simple thread pool implementations that are just a couple hundred lines of code, so I will try hooking up the code to one of those to see if we can drop the dependency. 

That's welcome, but if for some reasons it happens that other implementations don't perform as well as OMP, the libgomp isn't really an annoying dependency on both Linux and MSYS2.

illwieckz commented:
> I may have not tested yet the non-chunked code with the pragma indeed.

So I just somewhat tested it by using the amount of vertex as number of chunks, with a chunk of size 1.

I have hard time to see a difference on my workstation using 16-threads, but that's because I have other things running alongside, both (chunked or not) currently run at 400~410 fps.

On my 8-core laptop I see a small difference on the fact with the chunked implementation it sometime top at 183 fps while with the non-chunked implementation it doesn't top higher than 179 fps, I reproduced this multiple times.

illwieckz commented:
I unchunked the code.

If we want to investigate chunking, we can do it later, and if we do it we should do it in the template instead.

slipher commented:
I tried the compiler's built-in loop parallelization with `#pragma omp parallel for` -- see the `slipher/omp-for` branch. This gives me a measurable performance boost over the lambda-based dispatch.

slipher commented:
I decided not to bother trying the thread pool since the pragma-based approach with OMP actually seems the least intrusive: that way, there are no lambdas which would make the single-threaded version less efficient. And the amount of extra code is minimal. Also MSVC supposedly implements OMP, so I will try that later.

slipher commented:
> Though, you're not setting the thread count before running the loop, and I noticed that when not setting them right before running the loop, the amount of threads being used is unpredictable.

Changing the number of threads at runtime wouldn't work yet on my branch, but `Omp::Init` called on startup does, so it should work fine as long as you don't toggle the cvars.

slipher commented:
How are you determining that "the amount of threads being used is unpredictable"? It makes sense that turning on `r_smp` would throw off timing measurements by having another thread unpredictably running at the same time. So don't do that!

illwieckz commented:
By printing the output of `omp_get_num_threads()` and also by looking at the amount of busy threads in `htop`.

In my previous experiments I got very weird things, like `omp_get_num_threads()` returning `2` when I've set `16`, etc.

illwieckz commented:
I removed the `if (BUILD_SERVER)` etc. from the cmake file, I also removed the `if (NOT BUILD_CGAME)` etc. from it because I guess it would prevent to use OpenMP when building native games in the same cmake build as the engine, as we don't build native games in subprojects.

illwieckz commented:
I also removed the commits parallelizing loading stuff, that can be discussed later.

illwieckz commented:
> I decided not to bother trying the thread pool since the pragma-based approach with OMP actually seems the least intrusive.

Yes, if we can use OMP that would be very good, it's very easy to integrate in our code, and the code just build without problem when OMP is missing.

illwieckz commented:
Just as a test I commented out the `EnlistThreads()` calls and then the engine spawns 32 threads and the framerate is `1fps`.

illwieckz commented:
I added a log line in `/gfxinfo` reporting OpenMP support and the current amount of threads it will use.

slipher commented:
> I added the CMake stuff for MSVC.

It doesn't work yet on Windows, because `Omp.h` and `omp.h` are considered the same filename.

The other thing I am asking is to avoid adding the `libgomp` dependency to the server, which does not use it. So the new files should be placed in the renderer source list and the cvar should have an `r_` prefix to reflect the module which uses it.

I found that luckily the threads are not really started if they are not used (with GNU OpenMP at least), so it is not important to avoid requesting them when GPU vertex skinning is fully supported.

slipher commented on src/engine/framework/OmpSystem.cpp:
```diff
@@ -0,0 +1,104 @@
+{
+	ReadMaxThreads();
+	SetupThreads();
+	EnlistThreads();
```
This calls EnlistThreads twice in a row.

slipher commented on src/engine/framework/System.cpp:
```diff
@@ -829,6 +830,8 @@ static void SetCvarsWithInitFlag(cmdlineArgs_t& cmdlineArgs)
 // Initialize the engine
 static void Init(int argc, char** argv)
 {
+	Omp::Init();
```
Too early since cvars are not initialized.

illwieckz replied:
Where are Cvars initialized?

slipher commented on cmake/DaemonFlags.cmake:
```diff
@@ -267,6 +271,16 @@ if (MSVC)
     endif()
 
+	if (NOT NACL AND BUILD_CLIENT AND USE_OPENMP)
+		check_CXX_compiler_flag("/openmp" FLAG_OPENMP)
```
Flag checks don't work at all with MSVC so you just have to assume it's there.

slipher reviewed (changes requested):
Tested on MSVC. Had to make some small changes for it to compile due to `Error	C3016	'i': index variable in OpenMP 'for' statement must have signed integral type`. I'll push those to the branch. As hoped, I observed a speedup with `r_vboModels 0` vs. the 0.55.5 release. The threads cvar didn't work though (see code review comments).

illwieckz commented on src/engine/framework/OmpSystem.cpp:
```diff
@@ -0,0 +1,104 @@
+	if ( common_ompThreads.Get() )
+	{
+		ompThreads = common_ompThreads.Get();
+		return;
```
~~Cleaned-up.~~

illwieckz commented:
There is a remaining bug where changing the ompThreads value isn't applied unless I do `vid_restart`.

illwieckz commented:
> There is a remaining bug where changing the ompThreads value isn't applied unless I do `vid_restart`.

Now fixed.

illwieckz commented on src/engine/renderer/tr_init.cpp:
```diff
@@ -1057,6 +1058,21 @@ ScreenshotCmd screenshotPNGRegistration("screenshotPNG", ssFormat_t::SSF_PNG, "p
+			Log::Notice("%sUsing OpenMP parallelism with %d threads.", Color::ToString( Color::Green ), ompThreads );
+		}
+#else
+		Log::Notice("%sNot using OpenMP parallelism: unavailable.", Color::ToString( Color::Red ) );
```
I forgot this, now turned Yellow.

slipher reviewed (commented):
LGTM modulo comment

illwieckz commented on src.cmake:
```diff
@@ -304,16 +310,23 @@ set(CLIENTTESTLIST ${ENGINETESTLIST}
 )
 
+set(DUMMYAPPLIST
+    ${OMPLIST}
```
Fixed.

PS: Please comment on the problematic line itself, not the line above it.
Earlier in the day I fixed bad spacing on the `${OMPLIST}` line so I first believed your cache was obsolete, until I notice you talked about the `${COMMON_DIR}/Util.h` line below it (which GitHub doesn't show in main PR thread).

illwieckz replied:
If you comment on the last line not matching, GitHub will display both of them (and disclose the mismatch). If you comment on the first line not matching, GitHub will only display the first mismatching one (and hide the mismatch).
